swift/IceStorm/clock/Sources/Publisher/main.swift

```diff
@@ -113,27 +98,34 @@ func run() -> Int32 {
         let dateFormatter = DateFormatter()
         dateFormatter.dateFormat = "dd/MM/YYYY HH:mm:ss"
 
-        let t = DispatchSource.makeTimerSource()
-        t.schedule(deadline: .now(), repeating: .seconds(1))
-        t.setEventHandler {
-            do {
-                try clock.tick(dateFormatter.string(from: Date()))
-            } catch is CommunicatorDestroyedException {
-                t.suspend()
-                exit(0)
-            } catch {
-                t.suspend()
-                print("Error: \(error)\n")
-                communicator.destroy()
-                exit(1)
+        // Send a tick every second until cancelled
+        let task = Task {
+            while true {
+                do {
+                    try await clock.tick(dateFormatter.string(from: Date()))
+                } catch let error as Ice.LocalException {
+                    print("tick invocation failed with: \(error.message)")
+                }
+                try await Task.sleep(for: .seconds(1))
             }
         }
-        t.activate()
-        Dispatch.dispatchMain()
+
+        let signal = await ctrlCHandler.catchSignal()
+        print("Caught signal \(signal), exiting...")
+
+        task.cancel()
+        do {
+            try await task.value
+        } catch is CancellationError {
+            // expected
+        }
+
+        return 0
     } catch {
         print("Error: \(error)\n")
         return 1
     }
 }
 
-exit(run())
+let ctrlCHandler = Ice.CtrlCHandler()
+exit(await run(ctrlCHandler: ctrlCHandler))
```

Comment:
I am not fond of this "exit(run)" pattern, we should replace it.

With this pattern, it's critical to create the CtrlCHandler before the first await call.

Reply:
It's also unnecessary